src/fparser/two/tests/fortran2003/test_primary_r701.py

```diff
@@ -0,0 +1,74 @@
+# Copyright (c) 2019 Science and Technology Facilities Council
+
+# All rights reserved.
+
+# Modifications made as part of the fparser project are distributed
+# under the following license:
+
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+
+# 1. Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+
+# 2. Redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution.
+
+# 3. Neither the name of the copyright holder nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+'''Test Fortran 2003 rule R701 : primary type.
```

Comment:
R701 has 2 constraints that need to be tested or referenced (if they can't be tested for a valid reason) in this file ...
```
C701 (R701) The type-param-name shall be the name of a type parameter.
C702 (R701) The designator shall not be a whole assumed-size array
```

Reply:
Will add a comment to the test.

From what I can see though, these aren't parse rules and shouldn't be addressed *anywhere* in fparser... they are compiler rules that are applied on top of a parse tree once you have the full program context. Have I missed something on this?

Reply:
In these two cases I agree that fparser is not able to check these constraints in place. However, there are cases of constraints which require a modification to the available options so can be tested. For example ...

C201 (R208) An execution-part shall not contain an end-function-stmt, end-program-stmt, or endsubroutine-stmt.

It is an interesting angle that constraints could/should all be tested once we have the full program context (or at least as much as is available). I hadn't thought of it that way. That would certainly make for more readable code.

Anyway, the comments are fine. Thanks.